src/ast.rs

```diff
@@ -69,20 +67,40 @@ impl<'a> Program<'a> {
             .map(|index| &self.blocks[*index])
     }
 
-    fn build_block_index(blocks: &[Block]) -> Result<HashMap<u16, usize>, Error> {
+    fn build_program(mut blocks: Vec<Block>) -> Result<Program, Error> {
+        let mut program_blocks = Vec::new();
         let mut block_index = HashMap::new();
-        for (index, block) in blocks.iter().enumerate() {
+        let mut datum = Vec::new();
+        let mut index = 0;
+        for block in blocks.drain(..) {
             match block {
-                Block::Line { line_number, .. } => {
-                    if block_index.insert(*line_number, index).is_some() {
-                        return Err(Error::DuplicateLineNumber {
-                            line_number: *line_number,
-                        });
+                Block::Line {
+                    line_number,
+                    statement,
+                    statement_source,
+                } => match statement {
+                    Statement::Data(mut statement_datum) => {
+                        datum.append(&mut statement_datum);
```

Comment:
Why are we doing this during building and not interpretation? What did I miss?

Reply:
Because for the implementation of READ you need to know all data blocks in advance. Btw. for FOR we need something similar. We first need to add a small generated preamble before each FOR block, and this needs to be done before interpretation.